### DISCUSSION TITLE
Building Docker image to run Spark on Ray with RAPIDS

### DISCUSSION THREAD

**Discussion by chenya-zhang** [NVIDIA/spark-rapids#8062] [Q&A]
Hey there!

We are trying to experiment Spark on Ray with RAPIDS but not sure if RAPIDS can support this case. 

1) The example Dockerfile for spark-rapids k8s setup: https://nvidia.github.io/spark-rapids/docs/get-started/Dockerfile.cuda
2) The OSS project for Spark on Ray: https://github.com/oap-project/raydp

In 1), we find the below commands to copy items under `spark/`:
```
COPY spark/jars /opt/spark/jars
COPY spark/bin /opt/spark/bin
COPY spark/sbin /opt/spark/sbin
COPY spark/kubernetes/dockerfiles/spark/entrypoint.sh /opt/
COPY spark/examples /opt/spark/examples
COPY spark/kubernetes/tests /opt/spark/tests
COPY spark/data /opt/spark/data
```

In 2), if running `pip install raydp-nightly`, we can find `pyspark/`.
Under `pyspark/`, it has the below content.
```
/anaconda3/lib/python3.7/site-packages/pyspark# ls
__init__.py      context.py                java_gateway.py  rdd.py             statcounter.py
__pycache__      daemon.py                 join.py          rddsampler.py      status.py
_globals.py      data                      licenses         resource           storagelevel.py
_typing.pyi      examples                  ml               resultiterable.py  streaming
accumulators.py  files.py                  mllib            sbin               taskcontext.py
bin              find_spark_home.py        pandas           serializers.py     traceback_utils.py
broadcast.py     install.py                profiler.py      shell.py           util.py
cloudpickle      instrumentation_utils.py  py.typed         shuffle.py         version.py
conf.py          jars                      python           sql                worker.py
```

In this case, will there be concerns if we instead `COPY pyspark/jars /opt/pyspark/jars` or set `SPARK_HOME` to the existing `.../pyspark` installed by RayDP, 2) there is no `/kubernetes/dockerfiles/spark/entrypoint.sh` or `/kubernetes/tests` under `pyspark/`- I think they may not be required if we are able to launch Spark with RayDP on k8s.

Any suggestions or pointers would be very helpful, thanks!

**Comment by jlowe** [comment]
We're not familiar with the Spark on Ray project and have never tested in that environment.  As long as Spark on Ray still supports Spark plugins and Spark's GPU scheduling, it should work.

Most of the Dockerfile commands you're seeing are just to get Spark setup on Kubernetes in the container and isn't specific to the RAPIDS Accelerator.  The key parts in the Dockerfile for the RAPIDS Accelerator rather than generic Spark setup are primarily starting from an image that provides CUDA and these commands:
```
COPY rapids-4-spark_2.12-*.jar /opt/sparkRapidsPlugin
COPY getGpusResources.sh /opt/sparkRapidsPlugin
```

They get the RAPIDS Accelerator jar and the GPU scheduling script in the container so they can be referenced by the spark submit commands referenced at https://github.com/NVIDIA/spark-rapids/blob/branch-23.02/docs/get-started/getting-started-kubernetes.md#running-spark-applications-in-the-kubernetes-cluster.

In summary, once you get Spark setup in a container that provides CUDA and can schedule for GPUs, you should only need to ensure the RAPIDS Accelerator jar is in the container and the relevant startup configs are specified with the Spark job to get the RAPIDS Accelerator to work (again, assuming Spark on Ray supports Spark's GPU scheduling and Spark plugins).

**Comment by chenya-zhang** [comment]
Thanks @jlowe for the helpful suggestions. Will let you know.

**Comment by chenya-zhang** [comment]
Updates:

1. We built a docker image on top of `nvidia/cuda:11.5.0-devel-ubuntu18.04` and include `ray`, `raydp`, `java`, `rapids-4-spark_2.12-23.02.0.jar` and `getGpusResources.sh`. 
2. After deploying the Ray cluster with this image, we have a Ray worker running with 8 nvidia gpu in the cluster. 
3. If `exec` into the Ray worker and run `nvidia-smi`, we can see the below information:
```
 NVIDIA-SMI 470.161.03   Driver Version: 470.161.03   CUDA Version: N/A      |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  On   | 00000000:10:1C.0 Off |                    0 |
| N/A   22C    P0    50W / 400W |      0MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  On   | 00000000:10:1D.0 Off |                    0 |
| N/A   21C    P0    47W / 400W |      0MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled 
```
4. Please note that the CUDA version is `N/A`, not sure if it can be an issue.
5. We then use RayDP to launch a simple Spark session. We have tested RayDP is able to support the `configs` option before. 
```
spark = raydp.init_spark(app_name='raydp_pyspark_example',
                         num_executors=1,
                         executor_cores=1,
                         executor_memory='1GB',
                         configs={
                             "spark.executor.resource.gpu.amount": "1",
                             "spark.executor.resource.gpu.vendor": "nvidia.com",
                             "spark.executor.resource.gpu.discoveryScript": "/home/ray/anaconda3/lib/python3.7/site-packages/pyspark/jars/getGpusResources.sh"
                         })
```

Unfortunately, there we got the error that `No available node types can fulfill resource request {'gpu': 1.0, 'CPU': 1.0, 'memory': 1073741824.0}. Add suitable node types to this cluster to resolve this issue.`


If there is anything else we can do to verify the gpu resource, that would be super helpful. 
Thanks!

**Reply by jlowe** [reply]
> Unfortunately, there we got the error that No available node types can fulfill resource request {'gpu': 1.0, 'CPU': 1.0, 'memory': 1073741824.0}. Add suitable node types to this cluster to resolve this issue.

I don't think this message is coming from Spark or the RAPIDS Accelerator.  It indicates the cluster is not able to schedule for GPU resources properly implying it's a k8s or Ray setup issue.  Are you able to run a regular Ray job (no Spark-on-Ray) or a k8s job on the cluster that requests a GPU resource and is scheduled properly?  If the Ray cluster has not been setup with GPU support, that would explain the error message.  

Docs on how to setup k8s with GPUs is available at https://docs.nvidia.com/datacenter/cloud-native/kubernetes/install-k8s.html.  I'm not familiar with Ray and thus not sure what needs to be done to allow Ray to schedule for GPUs.

**Reply by chenya-zhang** [reply]
yep, we are able to run a regular Ray job on the cluster that requests a GPU resource. RayDP can support GPU resource scheduling via spark config and ray scheduler, e.g. https://github.com/oap-project/raydp/issues/119#issuecomment-812263187

**Reply by jlowe** [reply]
Are you able to schedule a Spark-on-Ray job that requests a GPU without using the RAPIDS Accelerator?  As I mentioned above, this error message is not coming from the RAPIDS Accelerator.  It seems to be coming from Ray.  That implies there's an issue with the configuration of Spark-on-Ray or with the Ray cluster itself.

**Reply by jlowe** [reply]
It would be good to verify the Ray cluster thinks the nodes have GPU resources (and how many GPU resources each node has available).  Again, I'm not familiar with Ray, but it appears you can get this from a `ray status` command or something similar.  It appears this is failing simply trying to schedule nodes to satisfy the executor resource requests, which is before the RAPIDS Accelerator has a chance to do anything.    Everything points to Spark requesting three resources, 'gpu', 'CPU', and 'memory', and the Ray cluster can't find a node with enough of those resources available.  Verifying Ray thinks those three resources are on the cluster and available seems to be the next step.

**Reply by chenya-zhang** [reply]
yes, I agree with you. I will verify the ray cluster status. 

btw. Is the CUDA version being `N/A` concerning?
More details can be found in 3# and 4# from https://github.com/NVIDIA/spark-rapids/discussions/8062#discussioncomment-5598196

**Reply by jlowe** [reply]
The CUDA version being N/A is a bit concerning, as it implies a CUDA runtime wasn't installed in the container.  For example, I suspect /usr/lib/x86_64-linux-gnu/libcuda.so.1 is probably missing which could explain how nvidia-smi cannot find CUDA.  It _might_ still work because we statically link the CUDA runtime in our native binaries and in theory that should only require the GPU driver, but I don't know if we've explicitly tested a setup where the CUDA runtime was missing.

**Reply by jlowe** [reply]
On second thought, nvidia-smi returning N/A for CUDA version may preclude CUDA programs from finding a CUDA-compatible device.     Did you base the image on one of the official nvidia/cuda images?  That is typically the easiest way to get a CUDA-compatible container, and spinning up one of those base images should show a proper CUDA version with nvidia-smi.

**Reply by chenya-zhang** [reply]
Hmm, we found that RayDP did not seem to be compatible with nvidia base images on GPU machines after multiple tries.

For example, the below is our simple Docker image (updated):
```
FROM nvidia/cuda:11.5.0-devel-ubuntu18.04

# https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212771
RUN apt-key adv --fetch-keys https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64/3bf863cc.pub

# Install Python and pip

RUN set -ex; \
  apt update && apt install curl gpg -y --force-yes 
RUN apt-get update && apt-get install -y --allow-unauthenticated python3.8-dev 
RUN apt-get install -y g++ python3-distutils
RUN update-alternatives --install /usr/bin/python python /usr/bin/python3.8 99  &&\
    update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.8 99
RUN curl https://bootstrap.pypa.io/get-pip.py -o /tmp/get-pip.py && python3.8 /tmp/get-pip.py
RUN mkdir -p /root

# Install Ray dependencies
RUN python3.8 -m  pip install ray raydp notebook
```

When deploying a `raycluster`, the worker k8s pod stuck at init `Started container wait-gcs-ready`. 
If checking the pod log of container `wait-gcs-ready`:
```
mesg: ttyname failed: Inappropriate ioctl for device
wait for GCS to be ready
wait for GCS to be ready
wait for GCS to be ready
```

If we remove installing `raydp`, there is no issue with `ray` and the k8s pod runs well. CUDA version can also be displayed.
If we install `ray` and `raydp` on a cpu image, there is no issue.

Curious, is it possible that there are special requirements for a library to be compatible with the nvidia base image?

**Reply by jlowe** [reply]
> is it possible that there are special requirements for a library to be compatible with the nvidia base image?

I do not know any special compatibility requirements.  If you have any questions about how the nvidia/cuda image was built, you can see the Dockerfiles at https://gitlab.com/nvidia/container-images/cuda/-/tree/master/dist/11.5.0/ubuntu1804.  I'd recommend comparing those Dockerfiles with the ones from your CPU base image, maybe there's something critical that's missing.

Alternatively, you could try taking your working CPU images with ray and raydp and try adding the CUDA runtime to them.

**Reply by jlowe** [reply]
Also I'm unable to build an image from the simple Dockerfile above.  The first problem I hit is that apt is unable to find python3.8-dev unless I do an apt-get update first.  After patching that the curl line fails because nvidia/cuda:11.5.0-devel-ubuntu18.04 doesn't have curl installed and nothing tries to install it before it's used.

**Reply by chenya-zhang** [reply]
@jlowe you are right! I moved a block installing internal stuffs before the python installation. 
I updated the Dockerfile above with these extra lines:
```
RUN set -ex; \
  apt update && apt install curl gpg -y --force-yes 
RUN apt-get update && apt-get install -y --allow-unauthenticated python3.8-dev 
```

**Reply by chenya-zhang** [reply]
> Alternatively, you could try taking your working CPU images with ray and raydp and try adding the CUDA runtime to them.

yep, this is what we did. 

our "cpu image" is very simple. 
we only installed `raydp` on top of a `ray` base image from https://hub.docker.com/r/rayproject/ray. 
it works well.. hmm.

```
FROM rayproject/ray:2.3.0

RUN pip install raydp
```

**Reply by jlowe** [reply]
So the image based on rayproject/ray, addig raydp, and then adding CUDA is not able to successfully start a raycluster, getting stuck at the "wait for GCS to be ready" step like the image based on nvidia/cuda?

**Reply by chenya-zhang** [reply]
oh I misunderstood you! 
I still used the CUDA base image but add `ray`, `raydp` from my cpu image to it.

To install the correct CUDA runtime, shall I follow the below link for `11.5.0-devel-ubuntu18.04`?

https://developer.nvidia.com/cuda-11-5-0-download-archive?target_os=Linux&target_arch=x86_64&Distribution=Ubuntu&target_version=18.04&target_type=deb_local

**Reply by jlowe** [reply]
Yes, you could follow the installation instructions for Debian packages either local or network in your Dockerfile.

**Reply by jlowe** [reply]
Any update on trying to add the CUDA toolkit to the functioning CPU image for Ray?  Curious if you were able to make this work.

**Reply by avinashahuja** [reply]
@chenya-zhang: Is this resolved? Can we mark this as closed? Would you be willing to share the steps that helped you resolve this?

**Reply by chenya-zhang** [reply]
@jlowe @avinashahuja hey folks, we are focusing on Spark Rapids GPU experiments at this moment. We can resume the discussion on this thread later.

**Comment by jlowe** [comment]
We were able to reproduce the `No available node types can fulfill resource request` problem, and I think we found a solution.  The `ray status` command gave a clue:
```
Resources
---------------------------------------------------------------
Usage:
 0.0/32.0 CPU
 0.0/2.0 GPU
 0B/128.00GiB memory
 0B/35.78GiB object_store_memory

Demands:
 {'CPU': 1.0, 'gpu': 1.0, 'memory': 1073741824.0}: 1+ pending tasks/actors
```

Note that the cluster is reporting `CPU`, `GPU`, and `memory` resources but receiving a request for `CPU`, `gpu`, and `memory` resources.  Note the case difference of GPU vs. gpu.  We were able to get the cluster to successfully allocate a GPU container with the following updates:
- changed the gpuGpusResources.sh script to use `\"GPU\"` instead of `\"gpu\"`
- changed the spark.executor.resource.gpu.* configs to corresponding spark.executor.resource.GPU.* configs
- changed spark.task.resource.gpu.amount to spark.task.resource.GPU.amount

These changes caused Spark to start asking for `GPU` resources instead of `gpu` resources.
